frontend/src/components/TimetableGrid.tsx

```diff
@@ -0,0 +1,238 @@
+
+export function TimetableGrid({
+  isVertical,
+  timetableDetailsSections,
+  handleUnitClick,
+  handleUnitDelete,
+}: {
+  isVertical: boolean;
+  timetableDetailsSections: {
+    id: string;
+    name: string;
+    roomTime: string[];
+    courseId: string;
+    type: string;
+    number: number;
+    instructors: string[];
+  }[];
+  handleUnitClick: (
+    e: {
+      id: string;
+      name: string;
+      courseId: string;
+      room: string;
+      code: string;
+      type: string;
+      number: number;
+      instructors: string[];
+    } | null,
+    event: React.MouseEvent,
+  ) => void;
+  handleUnitDelete: (
+    e: {
+      id: string;
+      name: string;
+      courseId: string;
+      room: string;
+      code: string;
+      type: string;
+      number: number;
+      instructors: string[];
+    } | null,
+  ) => void;
+}) {
+  const daysOfWeek = ["M", "T", "W", "Th", "F", "S"];
+
+  const timings = [
+    "8 - 9",
+    "9 - 10",
+    "10 - 11",
+    "11 - 12",
+    "12 - 1",
+    "1 - 2",
+    "2 - 3",
+    "3 - 4",
+    "4 - 5",
+    "5 - 6",
+    "6 - 7",
+  ];
+
+  const [displayRows, setDisplayRows] = useState(6);
+  const [displayCols, setDisplayCols] = useState(11);
+
+  const timetableGrid = useMemo(() => {
+    const grid: ({
+      id: string;
+      name: string;
+      courseId: string;
+      room: string;
+      code: string;
+      type: string;
+      number: number;
+      instructors: string[];
+    } | null)[] = Array(11 * 6);
+    for (let i = 0; i < grid.length; i++) {
+      grid[i] = null;
+    }
+    for (let i = 0; i < timetableDetailsSections.length; i++) {
+      for (let j = 0; j < timetableDetailsSections[i].roomTime.length; j++) {
+        const [code, room, day, hour] =
+          timetableDetailsSections[i].roomTime[j].split(":");
+        const remainder = daysOfWeek.indexOf(day);
+        const quotient = parseInt(hour) - 1;
+        grid[
+          isVertical ? remainder + quotient * 6 : quotient + remainder * 11
+        ] = {
+          id: timetableDetailsSections[i].id,
+          courseId: timetableDetailsSections[i].courseId,
+          room: room,
+          code: code,
+          type: timetableDetailsSections[i].type,
+          number: timetableDetailsSections[i].number,
+          instructors: timetableDetailsSections[i].instructors,
+          name: timetableDetailsSections[i].name,
+        };
+      }
+    }
+
+    const minDisplayCols = 9;
+    let nonNullColumns = 11;
+    for (let j = 10; j >= minDisplayCols; j--) {
+      let flag = true;
+      for (let i = 0; i < 6; i++) {
+        if ((isVertical ? grid[j * 6 + i] : grid[i * 11 + j]) !== null) {
+          flag = false;
+          break;
+        }
+      }
+      if (flag) {
+        nonNullColumns = j;
+      } else {
+        break;
+      }
+    }
+    setDisplayCols(nonNullColumns);
+
+    return grid;
+  }, [timetableDetailsSections, isVertical]);
+
+  return (
+    <div className="ml-4 flex w-full">
```

Comment:
```suggestion
    <div className="flex w-full">
```
Remove this margin, if the parent component wants to add spacing it can do so with padding anyways.